R/partial.R

```diff
@@ -43,45 +50,67 @@
 #' f()
 #' f()
 #'
-#' # You can override this by saying .lazy = FALSE
-#' f <- partial(runif, n = rpois(1, 5), .lazy = FALSE)
+#' # You can override this by unquoting:
+#' f <- partial(runif, n = !! rpois(1, 5))
+#' f
+#' f()
+#' f()
+#'
+#' # Partialized arguments can be spliced:
+#' args <- list(n = rpois(1, 5))
+#' f <- partial(runif, !!! args)
 #' f
 #' f()
 #' f()
 #'
-#' # This also means that partial works fine with functions that do
-#' # non-standard evaluation
+#' # partial() works fine with functions that do non-standard evaluation
 #' my_long_variable <- 1:10
 #' plot2 <- partial(plot, my_long_variable)
 #' plot2()
 #' plot2(runif(10), type = "l")
-partial <- function(...f, ..., .env = parent.frame(), .lazy = TRUE,
-                    .first = TRUE) {
-  stopifnot(is.function(...f))
+#'
+partial <- function(...f, ..., .first = TRUE, .env, .lazy) {
+  if (!missing(.env))
+    abort("`.env` is deprecated")
+  if (!missing(.lazy))
+    abort("`.lazy` is deprecated; unquote values to evalute them immediately")
 
-  if (.lazy) {
-    fcall <- substitute(...f(...))
-  } else {
-    fcall <- make_call(substitute(...f), .args = list(...))
-  }
+  f <- as_closure(...f)
+  args <- quos(...)
 
-  # Pass on ... from parent function
-  n <- length(fcall)
-  if (!.first && n > 1) {
-    tmp <- fcall[1]
-    tmp[[2]] <- quote(...)
-    tmp[seq(3, n + 1)] <- fcall[seq(2, n)]
-    names(tmp)[seq(3, n + 1)] <- names2(fcall)[seq(2, n)]
-    fcall <- tmp
-  } else {
-    fcall[[n + 1]] <- quote(...)
+  if (is_empty(args))
+    return(f)
+
+  fill_args <- function(call) {
+    vals <- lapply(args, eval_tidy)
```

Comment:
Is it right that these should get re-evaluated on every call? Or should they be lazily evaluated once?

Reply:
You are absolutely right. Only once. Thanks!

Reply:
I think `vals` should be (tidily) eagerly evaluated, even though this diverges from R‘s usual call semantics, as I think this fits better with the notion that `partial()` “fixes” default values. Also, the original function won't necessarily understand tidyeval. So if the partialized arguments are to comprehend tidyeval, they have to be tidyeval'd before being passed to `f`. Or am I missing something?

Reply:
Need to backtrack ... now I remember why I was setting `vals` to re-evaluate on every call: to indeed mimic R‘s call semantics, whereby each call (lazily) evaluates the arguments anew (except here the arguments are eagerly evaluated, for the reasons in the previous comment).

On the other hand, you can unquote to fix values. This is akin to, but of course not literally the same as, evaluating the arguments once. I believe this is consistent with the tidyverse practice of unquoting.

Example:
```R
# Evaluation of partialized arguments occurs with every call, so runs may differ in length
f <- partial(runif, n = rpois(1, 5))
f()  #> [1] 0.8655788 0.4433602 0.5719572 0.8810286 0.7277090
f()  #> [1] 0.1200088 0.9801471 0.8726577 0.2121102

# But the run length can be fixed by unquoting:
f <- partial(runif, n = !!rpois(1, 5))
f()  #> [1] 0.3450202 0.4374072 0.2143421 0.3036093
f()  #> [1] 0.5108066 0.4283015 0.7340954 0.3224990
```

Reply:
Oh of course — I wish we could describe the semantics of `partial()` better so that it's more clear to us what the desired behaviour is. My main concern with making any changes to `partial()` is that we'll accidentally change the API just because we don't remember what the desired semantics are.

Reply:
I think `.lazy` should not be made defunct but soft-deprecated. The default `.lazy = TRUE` would use the new behaviour which is consistent, and a new code path for the `FALSE` case should `eval_tidy()` once and for all.

Reply:
Hmm, I see your point. Maybe it is indeed better to enable quasiquotation but otherwise leave the behavior of `partial()`, and its argument signature, unchanged.

If you are willing to import the [nofrills package](https://github.com/egnha/nofrills) into purrr, then a `partial()` that satisfies these requirements can be quite succinct:
```R
partial <- function(...f, ..., .env = parent.frame(), .lazy = TRUE, .first = TRUE) {
  f <- substitute(...f)
  args <- if (.lazy) exprs(...) else lapply(quos(...), eval_tidy)
  if (.first)
    nofrills::fn(... ~ (!! f)(!!! args, ...), ..env = .env)
  else
    nofrills::fn(... ~ (!! f)(..., !!! args), ..env = .env)
}
```

Reply:
I think using quasiquotation to opt into eager behaviour is an interesting idea, just that the deprecation needs to happen incrementally. By the way if you make it an error to use an argument, this means it's defunct not deprecated as the latter only implies a warning. In general we try to soft-deprecate first without any warning (in the future testthat will send a warning for soft-deprecated functions but only when testing interactively).